data-analytics/dataflow-elasticsearch-indexer/ElasticIndexer/src/main/java/com/google/cloud/pso/utils/PublishProducts.java

```diff
@@ -0,0 +1,262 @@
+     */
+    public static void main(String[] args) {
+        PublishProductsOptions options =
+                PipelineOptionsFactory
+                        .fromArgs(args)
+                        .withValidation()
+                        .as(PublishProductsOptions.class);
+
+        run(options);
+    }
+
+    /**
+     * Runs the pipeline with the supplied options
+     *
+     * @param options The execution parameters to the pipeline.
+     * @return The result of the pipeline execution.
+     */
+    private static PipelineResult run(PublishProductsOptions options) {
+        // Create the pipeline
+        Pipeline pipeline = Pipeline.create(options);
+
+        /**
+         * Steps: 1) Create a filepattern based on the input GCS path.
+         *        2) Match the file pattern and read the matched patterns.
+         *        3) Read files, one per {@link ParDo} as the json messages
+         *           are not line delimited. The {@link DoFn} implementation
+         *           generates a separate json message for every product.
+         *        4) Read the product jsons and generate {@link Mutation}
+         *           using {@link JsonToMutationFn}
+         *        5) Write to BigTable
+         *        6) Publish to a PubSub topic.
+         */
+        PCollection<String> jsons =
+                pipeline
+                        //1) Create a filepattern based on the input GCS path.
+                        .apply("Create Filepattern",
+                                Create.of(options.getInput()))
+
+                        // 2) Match the file pattern and read the
+                        //    matched patterns.
+                        .apply(FileIO.matchAll()
+                                .withEmptyMatchTreatment(
+                                        EmptyMatchTreatment.DISALLOW))
+                        .apply(FileIO.readMatches()
+                                .withCompression(Compression.AUTO))
+
+                        // 3) Read files, one per {@link ParDo} as the json
+                        //    messages are not line delimited.
+                        //    The {@link DoFn} implementation generates a
+                        //    separate json message for every product.
+                        .apply("Parse and Extract",
+                                ParDo.of(
+                                        new DoFn<FileIO.ReadableFile, String>() {
+                                            @ProcessElement
+                                            public void processElement(ProcessContext context)
+                                                    throws IOException {
+                                                FileIO.ReadableFile readableFile =
+                                                        context.element();
+                                                String raw = readableFile
+                                                        .readFullyAsUTF8String();
```

Comment:
This will cause memory issues as file sizes grow. It would be better to process the file as a stream to make this solution more scalable.

Reply:
The reason I used this was because this helper class is only meant to be used to demo this pipeline by publishing a very specific dataset available at https://github.com/BestBuyAPIs/open-data-set/products.json.

The products.json data available in that open data set is unfortunately not new line delimited documents. I could potentially use Jackson's JsonParser to process this as a stream. However, I do not anticipate this particular class to be used for anything beyond a demo. In a real use case, only the core pipeline (IndexerMain) would be used.